### PR TITLE
linear fst with self loops

### PR DESCRIPTION
**Pull request by glynpu**
**What is this pr for?**

The new added linear_fst_with_self_loops is mainly solving crash with command.
```
python3 ./conformer_ctc/decode.py --exp-dir pruned_transducer_stateless_smooth/quandongtest_ctc_attnetion1-exp --lang-dir data/lang_bpe_500 --lm-dir data/lm --epoch 999 --avg 1 --max-duration 20  \
--method nbest-rescoring
```

and related logs are:
```
Traceback (most recent call last):                                                                                                                                          
  File "./conformer_ctc/decode.py", line 819, in <module>                                                                                                                   
    main()                                                                                                                                                                  
  File "/home/storage23/wangquandong/tools/k2_rnnt/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context                                   
    return func(*args, **kwargs)                                                                                                                                            
  File "./conformer_ctc/decode.py", line 794, in main                                                                                                                       
    results_dict = decode_dataset(                                                                                                                                          
  File "./conformer_ctc/decode.py", line 529, in decode_dataset                                                                                                             
    hyps_dict = decode_one_batch(                                                                                                                                           
  File "./conformer_ctc/decode.py", line 406, in decode_one_batch                                                                                                           
    best_path_dict = rescore_with_n_best_list(                                                                                                                              
  File "/home/storage23/wangquandong/experiment/open_source/icefall/icefall/decode.py", line 694, in rescore_with_n_best_list                                               
    nbest = nbest.intersect(G)                                                                                                                                              
  File "/home/storage23/wangquandong/experiment/open_source/icefall/icefall/decode.py", line 372, in intersect                                                              
    one_best = k2.invert(one_best)                                                                                                                                          
  File "/home/storage23/wangquandong/tools/k2/k2/python/k2/fsa_algo.py", line 780, in invert                                                                                
    if isinstance(fsa.aux_labels, torch.Tensor):                                                                                                                            
  File "/home/storage23/wangquandong/tools/k2/k2/python/k2/fsa.py", line 490, in __getattr__                                                                                
    raise AttributeError(f'Unknown attribute {name}')                                                                                                                       
AttributeError: Unknown attribute aux_labels 
```

**What's the reason for the crash?**
```
1. k2.invert(one_best) requires that the input one_best both have "labels" and "aux_labels".
2. the aux_labels of G are removed.
3. word_fsa indeed has aux_labes, but that are removed by:
word_fsa_with_epsilon_loops = k2.linear_fsa_with_self_loops(word_fsa)

So, it raises "AttributeError: Unknown attribute aux_labels "

```
related code
https://github.com/k2-fsa/icefall/blob/2f75236c05d38fb52bceb14e77ac104952e856cc/icefall/decode.py#L338-L340

### CODE REVIEW

**Comment by glynpu** [comment]
**Why use a cpu version of k2.remove_epsilon to generate expected out in unitest?**
(Also why suggest to use k2.connect after using a "GPU" k2.remove_epsilon in the comment.)
related code:
```
expected_fst = k2.add_epsilon_self_loops(
                    k2.remove_epsilon(src_fst.to("cpu"))
                ).to(device)
```

CPU and GPU version of k2.remove_epsilon are slight different. Here are some comparisons:
image

image



related code:
```
s1 = '''                                                                                                                                                            
          0 1 1 10 -0.1                                                                                                                                                     
          1 2 2 20 -0.2                                                                                                                                                     
          2 3 0 0 -0.5                                                                                                                                                      
          3 4 0 40 -0.1                                                                                                                                                     
          4 5 0 50 -0.1                                                                                                                                                     
          5 6 0 0 -0.1                                                                                                                                                      
          6 7 -1 0 -0.1                                                                                                                                                     
          7                                                                                                                                                                 
        '''
fst1 = k2.Fsa.from_str(s1, acceptor=False)
k2.to_dot(fst1)
# remove_epsilon
k2.to_dot(k2.remove_epsilon(fst1.to('cpu')))
k2.to_dot(k2.connect(k2.remove_epsilon(fst1.to('cuda'))))

# remove_epsilon_and_add_self_loops
k2.to_dot(k2.linear_fst_with_self_loops(fst1))
k2.to_dot(k2.remove_epsilon_and_add_self_loops(fst1.to('cpu')))
k2.to_dot(k2.remove_epsilon_and_add_self_loops(fst1.to('cuda')))
k2.to_dot(k2.connect(k2.remove_epsilon_and_add_self_loops(fst1.to('cuda'))))
```

File: k2/python/k2/fsa_algo.py [hunk 1]
```diff
@@ -131,6 +131,91 @@ def linear_fst(labels: Union[List[int], List[List[int]]],
...
+    #         new_arc_flag[i] = 0
+    #
+    # new_arc_flag only contains 0 or 1.
+    # 1 means:
+    # create a new dest_arc in out_fst for current src_arc.
+    # 0 means:
+    # current src_arc will share the dest_arc of previous src_arc.
+    new_arc_flag[1:] = non_epsilon_positions[0:-1]
+
+    dest_arc_row_ids = torch.cumsum(new_arc_flag, dim=0) - 1
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:175]: Could you explain why it can be used as `row_ids`?

Comment (glynpu) [reply k2/python/k2/fsa_algo.py:175]: As in the comment:
    # new_arc_flag only contains 0 or 1.
    # 1 means:
    # create a new dest_arc in out_fst for current src_arc.
    # 0 means:
    # current src_arc will share the dest_arc of previous src_arc.
    
    If current src_arc has it's own dest_arc, then it should has its own row_id.
    If current src_arc share the dest_arc of previous src_arc, then it also share the row_id.
    
    So torch.cumsum(new_arc_flag) contains the same information of row_id, except the offset 1.

File: k2/python/tests/linear_fst_with_self_loops_test.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline k2/python/tests/linear_fst_with_self_loops_test.py:1]: Please also add a line to `CMakeLists.txt` below
https://github.com/k2-fsa/k2/blob/1c396b9aee094cd56867dc5c5bd2fb40074ed126/k2/python/tests/CMakeLists.txt#L55

File: k2/python/k2/__init__.py [hunk 3]
```diff
@@ -66,6 +66,7 @@
 from .fsa_algo import linear_fsa
 from .fsa_algo import linear_fsa_with_self_loops
 from .fsa_algo import linear_fst
+from .fsa_algo import linear_fst_with_self_loops
```
Comment (csukuangfj) [inline k2/python/k2/__init__.py:69]: Please also make a change to icefall after merging this PR.

https://github.com/k2-fsa/icefall/blob/master/icefall/decode.py#L338
```python3
word_fsa_with_epsilon_loops = k2.linear_fsa_with_self_loops(word_fsa)
```

File: k2/python/k2/fsa_algo.py [hunk 4]
```diff
@@ -131,6 +131,101 @@ def linear_fst(labels: Union[List[int], List[List[int]]],
...
+    new_arc_flag[1:] = non_epsilon_positions[0:-1]
+
+    # Take fst1 in k2/python/tests/linear_fst_with_self_loops_test
+    # as an example.
+    # labels:       [1 2 0 0 0 0 -1]
+    # new_arc_flag: [1 1 1 0 0 0 0]
+    # torch.cumsum: [1 2 3 3 3 3 3]
+    # -1          : [0 1 2 2 2 2 2]
+    # The final result is the row_id of a dest_arc that
+    # eash aux_label/score belongs to.
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:183]:
```suggestion
    # each aux_label/score belongs to.
```

File: k2/python/k2/fsa_algo.py [hunk 5]
```diff
@@ -131,6 +131,101 @@ def linear_fst(labels: Union[List[int], List[List[int]]],
...
+    dest_arc_row_ids = torch.cumsum(new_arc_flag, dim=0, dtype=torch.int32) - 1
+
+    # Some dest_arc may correspond to a sequence of source arcs.
+    # So the aux_label on these kind of dest_arc may contain multi tokens.
```
Comment (csukuangfj) [inline k2/python/k2/fsa_algo.py:187]:
```suggestion
    # So the aux_label on these kinds of dest_arc may contain multi tokens.
```

**Review by csukuangfj — APPROVED** [review]
Thanks1

**Pull request merged**